CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+                .Setup(database => database.ScriptEvaluate(It.IsAny<byte[]>(), It.IsAny<RedisKey[]>(),
+                    It.Is<RedisValue[]>(values => values != null), It.IsAny<CommandFlags>()))
+                .Callback<byte[], RedisKey[], RedisValue[], CommandFlags>(Put);
+
+            // Get
+            _databaseMock
+                .Setup(database => database.ScriptEvaluate(It.IsAny<byte[]>(), It.IsAny<RedisKey[]>(),
+                    It.Is<RedisValue[]>(values => values == null), It.IsAny<CommandFlags>()))
+                .Returns<byte[], RedisKey[], RedisValue[], CommandFlags>((hash, keys, values, flags) =>
+                    RedisResult.Create(_fauxDatabase[keys[0]].Values));
+
+            // Remove
+            _databaseMock
+                .Setup(database => database.KeyDelete(It.IsAny<RedisKey>(), It.IsAny<CommandFlags>()))
+                .Callback<RedisKey, CommandFlags>((key, flags) => _fauxDatabase.Remove(key));
+        }
+
+        private void SetupConnectionMock()
+        {
+            _connectionMock.SetupGet(connection => connection.Configuration).Returns("connectionString");
+            _connectionMock.Setup(connection => connection.GetEndPoints(It.IsAny<bool>())).Returns(new EndPoint[1]);
+            _connectionMock.Setup(connection => connection.GetServer(It.IsAny<EndPoint>(), It.IsAny<object>()))
+                .Returns(_serverMock.Object);
+            _connectionMock.Setup(connection => connection.GetDatabase(It.IsAny<int>(), It.IsAny<object>()))
+                .Returns(_databaseMock.Object);
+        }
+
+        private void SetupCache()
+        {
+            double cleanupInterval = 100;
+            var cleanupTimer = new DefaultTimer(cleanupInterval);
+            _timeToLive = TimeSpan.FromSeconds(1);
+
+            _cache = CacheFactory.Build<DummyModel>(part => part
+                .WithProtoBufSerializer()
+                .WithSelfCleaningRedisConfiguration(_connectionMock.Object, cleanupTimer, _timeToLive,
+                    out string configurationKey)
+                .WithSelfCleaningRedisCacheHandle(configurationKey));
+
+            _onRemoveByHandleInvocations = new List<(CacheItemRemovedEventArgs, DateTime)>();
+            _cache.OnRemoveByHandle += (sender, args) => _onRemoveByHandleInvocations.Add((args, DateTime.Now));
+
+            _cache.CacheHandles.OfType<IStartable>().Single().Start();
+        }
+
+        #endregion
+
+        [TearDown]
+        public void TearDown()
+        {
+            _fauxDatabase.Clear();
+            _onRemoveByHandleInvocations.Clear();
+        }
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            _cache.Dispose();
+        }
+
+        [TestCase(1, false)]
+        [TestCase(2, true)]
+        [TestCase(3, false)]
+        [TestCase(4, true)]
+        [TestCase(5, false)]
+        public void Put_AddSomeAmountOfKeys_OnRemoveByHandleCalledAfterTimeToLiveWithCorrectValueAndReason(
```

Comment:
I think it will be nice to have all of the tests under `#region tests`. But apparently Im the only one in the team that likes regions, so you dont have to

Reply:
lol
Considering the tests are the main point of this class I don't think a region would fit for them